doc/generated/man/klatency.1
'\" t
.\"     Title: klatency
.\"    Author: [see the "AUTHOR" section]
.\" Generator: DocBook XSL Stylesheets v1.78.1 <http://docbook.sf.net/>
.\"      Date: 2008/04/19
.\"    Manual: Xenomai Manual
.\"    Source: Xenomai 2.6.5
.\"  Language: English
.\"
.TH "KLATENCY" "1" "2008/04/19" "Xenomai 2\&.6\&.5" "Xenomai Manual"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
klatency \- Xenomai kernel latency test
.SH "SYNOPSIS"
.sp
\fBklatency\fR [ options ]
.SH "DESCRIPTION"
.sp
\fBklatency\fR is part of the Xenomai test suite\&. It is a latency measurement tool dedicated to kernel\-based activities\&. The system must run a suitable Xenomai enabled kernel with the respective module (klat_pipe)\&.
.SH "OPTIONS"
.sp
\fBklatency\fR accepts the following options
.PP
\fB\-l <data\-lines per header>\fR
.RS 4
default=21, 0 to suppress headers
.RE
.PP
\fB\-T <test_duration_seconds>\fR
.RS 4
default=0, so ^C to end
.RE
.PP
\fB\-q\fR
.RS 4
suppresses RTD, RTH lines if \-T is used
.RE
.PP
\fB\-P <rt_pipe_no>\fR
.RS 4
number of testing pipe, default=auto
.RE
.SH "AUTHOR"
.sp
\fBklatency\fR was written by Philippe Gerum\&. This man page was written by Roland Stigge\&.


ChangeLog
2009-03-23  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/skins/vxworks/semLib.c (sem_destroy_internal) : Add missing
	rescheduling. Fix sem_read_proc().

2009-03-19  Philippe Gerum  <rpm@xenomai.org>

	* include/asm-generic/wrappers.h, src/nucleus/shadow.c: Fix
	generic bits for 2.6.29.

2009-03-18  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/skins/native/pipe.c: Fix streaming mode.

2009-03-18  Jan Kiszka  <jan.kiszka@siemens.com>

	* ksrc/nucleus/shadow.c: Trace syscall exits.

	* patching file include/asm-generic/syscall.h,
	ksrc/nucleus/shadow.c: Add xn_sys_current_info.

	* src/rtdk/rt_print.c: suspend/resume printer thread.

	* ksrc/skins/native/syscall.c, src/skins/native/task.c: Support
	T_WARNSW on task creation.

	* src/skins/native/task.c: Avoid lazy non-TLS rt_task_self.

	* src/skins/native/task.c: Silent migration before self-deletion.

	* src/skins, ksrc/nucleus/shadow.c: Rework initial Linux
	schedparam setting.

	* ksrc/nucleus, include/asm-generic, ksrc/arch/generic: Exclude
	CPUs from Xenomai use.

	* ksrc/nucleus/shadow.c: Cleanup taskexit hook.

	* ksrc/nucleus/shadow.c: Suppress switch warning upon signal
	receipt when the process is being debugged.

	* ksrc/nucleus/sched.c: Allow to change the watchdog timeout via a
	module parameter.

	* ksrc/arch/x86/usercopy_{32, 64}.c: Fix miscompilation of
	__do_strncpy_from_user.

	* src/rtdk, include/rtdk.h: Add assert_context helpers.

2009-03-18  Jan Kiszka  <jan.kiszka@siemens.com>

	* ksrc/skins/posix/syscall.c: Do not return -EFAULT on zero-length
	names. Instead, return -EINVAL when trying to create objects with
	empty names.

2009-03-13  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/skins/vrtx/mx.c: Streamline mutex support.

2009-03-12  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/skins/vrtx/task.c, src/skins/vrtx/task.c: Fix task creation.

2009-03-12  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/nucleus/shadow.c, ksrc/nucleus/synch.c: Introduce debug
	option to detect mutexes held in relaxed sections.

2009-03-07  Philippe Gerum  <rpm@xenomai.org>

	* src/skins/posix/shm.c (__wrap_munmap): Forward to glibc upon EBADF.

	* ksrc/nucleus/Config.in, ksrc/arch/{powerpc, x86}/Config.in: Add
	missing SYS_STACKPOOLSZ switch.

2009-02-26  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/arch/powerpc/patches: Update Adeos support to
	2.6.24.6-powerpc-DENX-2.4-03 (compatible with MPC5121-ADS
	hardware, see ksrc/arch/powerpc/patches/README).

2009-02-25  Jan Kiszka  <jan.kiszka@siemens.com>

	* ksrc/arch/x86/nmi.c: Rework NMI watchdog support.
	
2008-02-24  Stephane Fillod  <stephane.fillod@thomson.net>

	* src/skins/posix/shm.c (__wrap_mmap64): Make real call depend on
	configure detection (HAVE_MMAP64).
	
2009-02-24  Yi Li <liyi.dev@gmail.com>

	* configure.in: Remove -mfdpic switch. Set by default by
	bfin-linux-uclibc-gcc, and may cause error with bfin-uclinux-gcc.

2009-02-24  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/arch/{arm, blackfin, powerpc}/hal.c (rthal_timer_request):
	Port x86 fix of race window.

2009-02-24  Jan Kiszka  <jan.kiszka@siemens.com>

	* ksrc/arch/x86/hal-common.c (rthal_timer_request): Fix race.

	* ksrc/nucleus/shadow.c (gatekeeper_thread): Do not refer to
	uninit sched slots.

2009-02-23  Paul Corner <paul_c@tuxcnc.org>

	* debian/prepare-patch.sh: Fix patch generation rules.

2009-02-23  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/arch/x86/patches: Upgrade to 2.6.27.19-x86-2.2-06,
	2.6.28.7-x86-2.2-06.

2009-02-17  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/arch/x86/patches: Upgrade to 2.6.27.13-x86-2.2-05.

	* ksrc/arch/powerpc/patches: Upgrade to
	2.6.28.5-powerpc-DENX-2.4-04.

2009-02-13  Gilles Chanteperdrix  <gilles.chanteperdrix@xenomai.org>

	* src/skins/posix/shm.c (__wrap_mmap, __wrap_mmap64): Map only
	part of the underlying nucleus heap.

	* ksrc/nucleus/heap.c (xnheap_mmap): Allow partial mapping in
	user-space of a nucleus heap.

	* ksrc/nucleus/sched.c, include/nucleus/sched.h: Introduce
	sched->cpu field member.

2009-02-12  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/nucleus/pod.c (xnpod_start_thread): Run start hook for
	user-space threads as well.

2009-02-11  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/skins/posix/shm.c (ftruncate): Remove GFP_USER hint.

2009-02-07  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/arch/blackfin, include/asm-blackfin: Prepare for 2.6.28.

	* ksrc/arch/blackfin/patches: Upgrade to 2.6.28-1.9-00 (GIT).

	* ksrc/nucleus, include/nucleus: Introduce host tick deferral.

2009-02-02  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/skins/native/syscall.c (rt_queue_write): Fix memory leak.

2009-02-01  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/nucleus/heap, include/asm-generic/system.h, ksrc/skins: Fix
	auto-cleanup for shared heaps.

2009-01-31  Gilles Chanteperdrix  <gilles.chanteperdrix@xenomai.org>

	* ksrc/drivers/testing/switchtest.c,
	  src/testsuite/switchtest/switchtest.c: Add stress mode to
	switchtest.

2009-01-30  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/arch/ia64, include/asm-ia64: Itanium port discontinued.

2009-01-30  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/arch/x86/patches: Upgrade to 2.6.27.13-x86-2.2-04,
	2.6.28.2-x86-2.2-05.

2009-01-27  Gilles Chanteperdrix  <gilles.chanteperdrix@xenomai.org>

	* include/asm-x86/bits/pod_{32, 64}.h (xnarch_leave_root): Fix FPU
	backup clobbering.
	
2009-01-27  Philippe Gerum  <rpm@eclipse.xenomai.org>

	* ksrc/nucleus, include/nucleus, ksrc/skins/posix, include/posix,
	sim, include/asm-sim: Introduce POSIX sporadic scheduling (PSS)
	class.

2009-01-21  Gilles Chanteperdrix  <gilles.chanteperdrix@xenomai.org>

	* ksrc/arch/arm/patches: Upgrade to 2.6.28-arm-1.2-00.
	
2009-01-19  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/nucleus/pipe.c, ksrc/skins/native/pipe.c,
	ksrc/skins/rtai/fifo.c: Refactor callback management, fixing a
	severe latency issue, and a potential memory leak in both
	disconnect/release handlers.

	* ksrc/nucleus/pipe.c (xnpipe_disconnect, xnpipe_release):

src/testsuite/klatency/klatency.c
/*
 * Copyright (C) 2008 Gilles Chanteperdrix <gch@xenomai.org>
 *
 * Permission is hereby granted, free of charge, to any person obtaining
 * a copy of this software and associated documentation files (the
 * "Software"), to deal in the Software without restriction, including
 * without limitation the rights to use, copy, modify, merge, publish,
 * distribute, sublicense, and/or sell copies of the Software, and to
 * permit persons to whom the Software is furnished to do so, subject to
 * the following conditions:
 *
 * The above copyright notice and this permission notice shall be included
 * in all copies or substantial portions of the Software.
 *  
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
 * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
 * CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
 * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

#include <unistd.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <signal.h>
#include <getopt.h>
#include <time.h>
#include <errno.h>
#include <rtdm/rttesting.h>

struct pkt {
	struct rttst_tmbench_config config;
	struct rttst_interm_bench_res res;
};

long long period_ns = 0;
int test_duration = 0;		/* sec of testing, via -T <sec>, 0 is inf */
int data_lines = 21;		/* data lines per header line, -l <lines> to change */
int quiet = 0;			/* suppress printing of RTH, RTD lines when -T given */
int benchdev_no = -1;
int benchdev = -1;
int freeze_max;
int priority;

#define USER_TASK       0
#define KERNEL_TASK     1
#define TIMER_HANDLER   2

int test_mode = USER_TASK;
const char *test_mode_names[] = {
	"periodic user-mode task",
	"in-kernel periodic task",
	"in-kernel timer handler"
};

time_t test_start, test_end;	/* report test duration */
int test_loops = 0;		/* outer loop count */

int finished = 0;

void display(void)
{
	int err, n = 0, got_results = 0;
	time_t start, actual_duration;
	struct pkt result;

	time(&start);

	printf("warming up...\n");

	if (quiet)
		fprintf(stderr, "running quietly for %d seconds\n",
			test_duration);

	while (!finished) {
		long minj, gminj, maxj, gmaxj, avgj, goverrun;

		err = read(benchdev, &result, sizeof(result));
		if (err <= 0) {
			fprintf(stderr, "read: %d, errno: %d\n", err, errno);
			break;
		}

		got_results = 1;
		minj = result.res.last.min;
		gminj = result.res.overall.min;
		avgj = result.res.last.avg;
		maxj = result.res.last.max;
		gmaxj = result.res.overall.max;
		goverrun = result.res.overall.overruns;

		if (!quiet) {
			if (data_lines && (n++ % data_lines) == 0) {
				time_t now, dt;
				time(&now);
				dt = now - start;
				printf
					("RTT|  %.2ld:%.2ld:%.2ld  (%s, %Ld us period, "
					 "priority %d)\n", dt / 3600,
					 (dt / 60) % 60, dt % 60,
					 test_mode_names[test_mode],
					 period_ns / 1000, priority);
				printf("RTH|%12s|%12s|%12s|%8s|%12s|%12s\n",
				       "-----lat min", "-----lat avg",
				       "-----lat max", "-overrun",
				       "----lat best", "---lat worst");
			}

			printf("RTD|%12.3f|%12.3f|%12.3f|%8ld|%12.3f|%12.3f\n",
			       (double)minj / 1000,
			       (double)avgj / 1000,
			       (double)maxj / 1000,
			       goverrun,
			       (double)gminj / 1000, (double)gmaxj / 1000);
		}
	}

	time(&test_end);
	actual_duration = test_end - test_start;
	if (!test_duration)
		test_duration = actual_duration;

	if (got_results) {
		long gminj, gmaxj, gavgj, goverrun;

		gminj = result.res.overall.min;
		gmaxj = result.res.overall.max;
		goverrun = result.res.overall.overruns;
		gavgj = result.res.overall.avg
			/ ((result.res.overall.test_loops) > 1 ?
			   result.res.overall.test_loops : 2) - 1;

		printf("---|------------|------------|------------|--------|-------------------------\n"
		       "RTS|%12.3f|%12.3f|%12.3f|%8ld|    %.2ld:%.2ld:%.2ld/%.2d:%.2d:%.2d\n",
		       (double)gminj / 1000, (double)gavgj / 1000, (double)gmaxj / 1000,
		       goverrun, actual_duration / 3600, (actual_duration / 60) % 60,
		       actual_duration % 60, test_duration / 3600,
		       (test_duration / 60) % 60, test_duration % 60);

	}

	if (benchdev >= 0)
		close(benchdev);
}

void sighand(int sig __attribute__ ((unused)))
{
	finished = 1;
}

int main(int argc, char **argv)
{
	struct pkt pkt;
	int c;

	while ((c = getopt(argc, argv, "l:T:qP:")) != EOF)
		switch (c) {
		case 'l':

			data_lines = atoi(optarg);
			break;

		case 'T':

			test_duration = atoi(optarg);
			alarm(test_duration);
			break;

		case 'q':

			quiet = 1;
			break;

		case 'P':

			benchdev_no = atoi(optarg);
			break;

		default:

			fprintf(stderr, "usage: latency [options]\n"
				"  [-l <data-lines per header>] # default=21, 0 to suppress headers\n"
				"  [-T <test_duration_seconds>] # default=0, so ^C to end\n"
				"  [-q]                         # suppresses RTD, RTH lines if -T is used\n"
				"  [-P <rt_pipe_no>]            # number of testing pipe, default=auto\n");
			exit(2);
		}

	if (!test_duration && quiet) {
		fprintf(stderr,
			"latency: -q only works if -T has been given.\n");
		quiet = 0;
	}

	time(&test_start);

	signal(SIGINT, sighand);
	signal(SIGTERM, sighand);
	signal(SIGHUP, sighand);
	signal(SIGALRM, sighand);

	setlinebuf(stdout);

	if (benchdev_no == -1) {
		benchdev = open("/proc/xenomai/registry/native/pipes/klat_pipe",
				O_RDONLY);
		if (benchdev == -1) {
			perror("open(/proc/xenomai/registry/native/pipes/klat_pipe)");
			fprintf(stderr,
				"modprobe klat_mod or try the -P option?\n");
			exit(EXIT_FAILURE);
		}
	} else {
		char devname[64];
		snprintf(devname, sizeof(devname), "/dev/rtp%d", benchdev_no);
		benchdev = open(devname, O_RDONLY);
		if (benchdev == -1) {
			fprintf(stderr, "open(%s): %s\n",
				devname, strerror(errno));
			exit(EXIT_FAILURE);
		}
	}

	if (read(benchdev, &pkt, sizeof(pkt)) == -1) {
		perror("read");
		exit(EXIT_FAILURE);
	}

	test_mode = pkt.config.mode;
	priority = pkt.config.priority;
	period_ns = pkt.config.period;
	freeze_max = pkt.config.freeze_max;

	printf("== Sampling period: %Ld us\n"
	       "== Test mode: %s\n"
	       "== All results in microseconds\n",
	       period_ns / 1000, test_mode_names[test_mode]);

	display();

	return 0;
}
